### PR TITLE
Cache the perferred resources and namespaced resources

### PR DESCRIPTION
**Pull request by tossmilestone**
**What this PR does / why we need it**:

Fixes the following TODO:
https://github.com/kubernetes/kubernetes/blob/99fd98a893b49fbc32e933cf9cbd0e1eb8b9313a/staging/src/k8s.io/client-go/discovery/cached/memcache.go#L99-L101

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
None

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by tossmilestone** [comment]
@sttts, please help to review the PR if you have time. Thanks! :smile:

**Comment by tossmilestone** [comment]
/unassign @sttts 
/assign @liggitt please help to review the PR, thanks!
😄

**Comment by tossmilestone** [comment]
/assign @liggitt please help to review the PR, thanks!
😄

**Comment by tossmilestone** [comment]
/assign @liggitt
please help to review the PR, thanks!
😄

**Comment by smarterclayton** [comment]
@kubernetes/sig-api-machinery-pr-reviews

File: staging/src/k8s.io/client-go/discovery/cached/memcache.go [hunk 1]
```diff
@@ -96,18 +98,24 @@ func (d *memCacheClient) RESTClient() restclient.Interface {
...
-// TODO: Should this also be cached? The results seem more likely to be
-// inconsistent with ServerGroups and ServerResources given the requirement to
-// actively Invalidate.
 func (d *memCacheClient) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
-	return d.delegate.ServerPreferredResources()
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	if !d.cacheValid {
+		// If not cached, returns from the delegate.
+		return d.delegate.ServerPreferredResources()
```
Comment (sttts) [inline staging/src/k8s.io/client-go/discovery/cached/memcache.go]: The other methods return `ErrCacheEmpty` in this case. Why do we call the delegate here?

Comment (tossmilestone) [reply staging/src/k8s.io/client-go/discovery/cached/memcache.go]: You're right. It should return `ErrCacheEmpty` to have the same logic with other methods. I will update soon.

Comment (tossmilestone) [reply staging/src/k8s.io/client-go/discovery/cached/memcache.go]: Updated, PTAL @sttts

**Comment by sttts** [comment]
Please squash.

/approve

Leaving final lgtm to @caesarxuchao

**Comment by tossmilestone** [comment]
@sttts OK

File: staging/src/k8s.io/client-go/discovery/cached/memcache.go [hunk 2]
```diff
@@ -96,18 +98,22 @@ func (d *memCacheClient) RESTClient() restclient.Interface {
 	return d.delegate.RESTClient()
 }
 
-// TODO: Should this also be cached? The results seem more likely to be
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/discovery/cached/memcache.go]: Do you understand this TODO? What's the original concern?

This PR looks ok, but I'm not surew what this TODO tries to convey.

Comment (tossmilestone) [reply staging/src/k8s.io/client-go/discovery/cached/memcache.go]: This is my understanding. The TODO is considering the possible inconsistency between the resources and preferred resources, one from the cache, the other from the delegate. To cache them all could ensure the consistency. 
@caesarxuchao

Comment (sttts) [reply staging/src/k8s.io/client-go/discovery/cached/memcache.go]: Also if we cache only half of the data and the other half (preferred resources) retrieved in real-time, we could get inconsistencies.

Comment (caesarxuchao) [reply staging/src/k8s.io/client-go/discovery/cached/memcache.go]: Make sense.

**Comment by tossmilestone** [comment]
@lavalamp that's a bit change needs your approval, really appreciate if you can take a look.

**Comment by tossmilestone** [comment]
@caesarxuchao needs your lgtm again, thanks!

File: staging/src/k8s.io/client-go/discovery/cached/memcache.go [hunk 3]
```diff
@@ -162,7 +168,19 @@ func (d *memCacheClient) Invalidate() {
 		}
 	}
 
-	d.groupToServerResources, d.groupList = rl, gl
+	prl, err := d.delegate.ServerPreferredResources()
```
[the remark this answers is not in the document]
Comment (tossmilestone) [reply staging/src/k8s.io/client-go/discovery/cached/memcache.go]: @liggitt Seems your solution is better, I will close this PR then.

**Pull request closed without merging**